File: python/302.SmallestRectangleEnclosingBlack.py
class Solution(object):
	def minArea(self, image, x, y):
		"""
		:type image: List[List[str]]
		:type x: int
		:type y: int
		:rtype: int
			"""
		def first(lo, hi, check):
			while lo < hi:
				mid = (lo + hi) / 2
				if check(mid):
					hi = mid
				else:
					lo = mid + 1
			return lo
		top    = first(0, x,             lambda x: '1' in image[x])
		bottom = first(x, len(image),    lambda x: '1' not in image[x])
		left   = first(0, y,             lambda y: any(row[y] == '1' for row in image))
		right  = first(y, len(image[0]), lambda y: all(row[y] == '0' for row in image))
		return (bottom - top) * (right - left)

File: javascript/302.SmallestRectangleEnclosingBlack.js
/**
 * @param {character[][]} image
 * @param {number} x
 * @param {number} y
 * @return {number}
 */
var minArea = function(image, x, y) {
	top    = first(0, x,               (x) => image[x].includes('1'));
	bottom = first(x, image.length,    (x) => !image[x].includes('1'));
	left   = first(0, y,               (y) => image.some((row) => row[y] === '1'));
	right  = first(y, image[0].length, (y) => image.every((row) => row[y] !== '1'));
	return (bottom - top) * (right - left);
};
var first = function(lo, hi, check) {
	while (lo < hi) {
		let mid = Math.floor((lo + hi) / 2);
		if (check(mid)) hi = mid;
		else lo = mid + 1;
	}
	return lo;
};